fix(nlu): check ram usage

ierezell opened the pull request:
Added a small fix to check ram before launchin the nlu server as asked in issue botpress/v12#1154.

Commit:
NLU fix : check ram usage

Commit:
Fix(nlu) : Added condition on languages

JustusNBB commented on src/bp/lang-server/service/index.ts:
```diff
@@ -19,6 +25,14 @@ const FAST_TEXT_MODEL_REGEX = /^(\w+)\.(\w+)\.(\d+)\.bin$/i
+  return {
+    used: Number.parseFloat((os.freemem() / (1024 * 1024 * 1024)).toFixed(2)),
+    total: Number.parseFloat((os.totalmem() / (1024 * 1024 * 1024)).toFixed(2)),
+    prediction: (dims > 101 ? 4.0 : 2.0) * langNb
```
This line is the core logic of this PR.

Idea: Document prediction thresholds/factors more verbosely in code and possibly also in the warning itself. Afaik requirements vary from 1 to 6GB per language depending on the language and dimensions used (25, 100, 300).
E.g. create constants for the values "4.0" and "2.0" to be more clearly readable/adaptable to be tweaked - maybe 5.0 and 1.5 fit better?

Ierezell replied:
A next step could be to have a map per language & dims -> size. I will add something like that to be more precise because all models are not equals. 

The problem is that `os.freemem()` is giving the free memory (including the virtual one) so much less than the "real" available one.  Thus it will trigger the warning even if there is sometimes enough memory.

I'm still looking for a better solution to get the "real" ram available but node don't seem to do such Os things and we don't want to add a 50Mb library just for it.

franklevasseur commented on src/bp/lang-server/service/index.ts:
```diff
@@ -19,6 +25,14 @@ const FAST_TEXT_MODEL_REGEX = /^(\w+)\.(\w+)\.(\d+)\.bin$/i
 
+const checkEnoughRam = (langNb: number, dims: number): RamInfos => {
+  return {
+    used: Number.parseFloat((os.freemem() / (1024 * 1024 * 1024)).toFixed(2)),
```
checkout `bytes` package

allardy replied:
Actually, there's a method called bytesToString which will use the correct naming

franklevasseur commented on src/bp/lang-server/service/index.ts:
```diff
@@ -19,6 +25,14 @@ const FAST_TEXT_MODEL_REGEX = /^(\w+)\.(\w+)\.(\d+)\.bin$/i
+  return {
+    used: Number.parseFloat((os.freemem() / (1024 * 1024 * 1024)).toFixed(2)),
+    total: Number.parseFloat((os.totalmem() / (1024 * 1024 * 1024)).toFixed(2)),
+    prediction: (dims > 101 ? 4.0 : 2.0) * langNb
```
Extract `4.0` and `2.0` in constant. Something like `GB_PER_LANG`.

Ierezell replied:
I added a fonction based on a linear regression to get the size (more generic).

franklevasseur commented on src/bp/lang-server/service/index.ts:
```diff
@@ -41,6 +55,20 @@ export default class LanguageService {
+      const ramInfos = checkEnoughRam(languages.length, this.dim)
+
+      if (ramInfos.prediction < ramInfos.total - ramInfos.used) {
+        this.logger.info(`Lang server will take ~${ramInfos.prediction}Gb or more`)
```
There's no need to tell the user he has enough memory.

EFF replied:
Hmmm right, if the user has enough memory, it'll run, if not it'll crash. What we can do though is to log something if the user doesn't have enough mem. I suggest that required mem is computed with language model size + a buffer of let's say 20% of the model size because of all the caching. something like that :

```
requiredMem = modelSize + .2 * modelSize
if availableMem < requiredMem
  log "Not enough memory to start lang server, free mem and try again"
  process.exit(1)
```

franklevasseur reviewed (commented):
Hi there!

I'm not a fan of this solution. If the lang server needs more RAM when booting, it will crash right away anyway. 

The real problem occurs when another app like, for instance, botpress starts increasing its memory usage. The lang server can then suddenly crash silently even if botpress's memory usage then lowers down.

A real solution to this problem would be to know when the lang server starts lacking memory. This way we could either intervene so it doesn't happen or warn the user.

Again, I don't mind adding this logging. Maybe something like this would me more appropriate for all users:

`"Warning, this program can crash silently if it lacks free memory space. Make you sure you have enough RAM available."`

@EFF what do think?

EFF commented:
> Hi there!
> 
> I'm not a fan of this solution. If the lang server needs more RAM when booting, it will crash right away anyway.
> 
> The real problem occurs when another app like, for instance, botpress starts increasing its memory usage. The lang server can then suddenly crash silently even if botpress's memory usage then lowers down.
> 
> A real solution to this problem would be to know when the lang server starts lacking memory. This way we could either intervene so it doesn't happen or warn the user.
> 
> Again, I don't mind adding this logging. Maybe something like this would me more appropriate for all users:
> 
> `"Warning, this program can crash silently if it lacks free memory space. Make you sure you have enough RAM available."`
> 
> @EFF what do think?

I agree with this. Ideally we would warn the user when the server is close to lack mem

allardy commented:
We could add a check on startup and every 10 seconds to display a message if the "buffer" zone (of say 20%) is reached. This way we would have advanced warning and explanations if the server ends up crashing.

I would also push a bit further and make the process more resilient. Let's say you load 5 languages in memory. Maybe one of those language is not used often, so we could keep track of when a model was last queried, and unload it from memory temporarily, and it would be loaded the next time a query comes in.

It will probably not handle all scenarios, but could protect the server if there's a usage peak which goes down afterwards.

We could also add the Language server to the Monitoring/Alerting on the admin panel, and we could also store the logs in the database as we do for the botpress server...

Lots of possible updates to do.

Commit:
Merge branch 'master' of github.com:botpress/botpress into ps_warn_ram

Commit:
added interval

Ierezell commented:
I added an interval in the code which check every minute if there is enough free ram. 
However, the `os.freemem()` compute the free memory in a special way as said above : 

"""os.freemem() is giving the free memory (including the virtual one) so much less than the "real" available one. Thus it will trigger the warning even if there is sometimes enough memory."""

I lowered the threshold to check if there is less than 1% of the total memory free but we can adjust it. 

Also added a function to compute approximately the memory taken (with a small linear regression between dims and size).

Commit:
Merge branch 'master' into ps_warn_ram

EFF commented on src/bp/lang-server/service/index.ts:
```diff
@@ -3,22 +3,43 @@ import fs from 'fs'
 const BPE_MODEL_REGEX = /^(\w+)\.(\w+)\.bpe\.model$/i
 
+const getModelSize = (dims: number) => {
+  return (7.009615385 * dims + 913.7019231) / 1000
```
can you extract them in constants and name them properly so we know why you put those numbers (I'm sure they're not random)

Ierezell replied:
Indeed as said above I did a linear regression between embeddings dims and model size.

EFF commented on src/bp/lang-server/service/index.ts:
```diff
@@ -3,22 +3,43 @@ import fs from 'fs'
+
 const maxAgeCacheInMS = ms('24h')
 
+const MARGIN_PRED = 0.2 // Adds some % to the predicted ram usage
```
can you name this properly and remove the comment.
`MODEL_SAFETY_BUFFER` or something like this

Ierezell replied:
Done and did the same bellow for the ram threshold

EFF commented on src/bp/lang-server/service/index.ts:
```diff
@@ -3,22 +3,43 @@ import fs from 'fs'
+
+const checkEnoughRam = (langNb: number, dims: number): RamInfos => {
+  return {
+    free: Number.parseFloat((os.freemem() / (1024 * 1024 * 1024)).toFixed(2)),
```
we have the bytes lib for this

`bytes('1gb') === 1024 * 1024 * 1024`

EFF commented on src/bp/lang-server/service/index.ts:
```diff
@@ -41,10 +62,34 @@ export default class LanguageService {
     this._models = this._getModels()
     const languages = Object.keys(this._models)
 
+    if (languages.length > 0) {
+      const ramInfos = checkEnoughRam(languages.length, this.dim)
+
+      if (ramInfos.prediction + ramInfos.prediction * MARGIN_PRED > ramInfos.free) {
+        this.logger.warn(
+          'Warning, this program can crash silently if it lacks free memory space. Make you sure you have enough RAM available.'
+        )
```
this is duplicated code (here and in the interval. please extract this as a private method

Ierezell replied:
Added private `warnRam` method

EFF commented on src/bp/lang-server/service/index.ts:
```diff
@@ -41,10 +62,34 @@ export default class LanguageService {
+        this.logger.warn(
+          'Warning, this program can crash silently if it lacks free memory space. Make you sure you have enough RAM available.'
+        )
+        this.logger.warn(
+          `Lang server will take ${ramInfos.prediction}Gb and you have approximately ${ramInfos.free}Gb left on ${ramInfos.total}`
+        )
+      }
```
this warning is useless in my opinion. This should not be a warning, it's really informative, change .warn to .info please

EFF commented on src/bp/lang-server/service/index.ts:
```diff
@@ -41,10 +62,34 @@ export default class LanguageService {
+            ramInfos.free}/${ramInfos.total})`
+        )
+      }
+    }, 60000)
```
use the ms package please

`ms('1m')`

Ierezell replied:
needed a comment to finish the review on vscode, useful extension but i still don't get everything

EFF commented on src/bp/lang-server/service/index.ts:
```diff
@@ -3,22 +3,43 @@ import fs from 'fs'
+  return (7.009615385 * dims + 913.7019231) / 1000
+}
+
+const checkEnoughRam = (langNb: number, dims: number): RamInfos => {
```
There is actually a private method that does something similar `_getResourceConsumptionInfo` did you check it ? can we adapt it so it suits our need ?

Ierezell replied:
It could... but it's different... It uses process.memoryUsage() to get the Resident Set allocated by node (V8 etc...) however, it cannot know memory used by bindings or other programs

EFF reviewed (commented):
A few more comments

Commit:
Merge branch 'master' of github.com:botpress/botpress into ps_warn_ram

Commit:
Merge branch 'ps_warn_ram' of github.com:botpress/botpress into ps_warn_ram

Ierezell commented on src/bp/lang-server/service/index.ts:
```diff
@@ -3,22 +3,43 @@ import fs from 'fs'
+  return (7.009615385 * dims + 913.7019231) / 1000
+}
+
+const checkEnoughRam = (langNb: number, dims: number): RamInfos => {
```
It could... but it's different... It uses process.memoryUsage() to get the Resident Set allocated by node (V8 etc...) however, it cannot know memory used by bindings or other programs

Commit:
Fix(NLU) : Fixed eff comments

Commit:
Fix(NLU) : fixed typo

franklevasseur commented on src/bp/lang-server/service/index.ts:
```diff
@@ -41,10 +72,29 @@ export default class LanguageService {
+    if (languages.length > 0) {
+      const ramInfos = checkEnoughRam(languages.length, this.dim)
+
+      if (ramInfos.prediction + ramInfos.prediction * MODEL_SAFETY_BUFFER > ramInfos.free) {
```
mise en évidence simple:

ramInfos * (1 + MODEL_SAFETY_BUFFER) > ramInfos.free

franklevasseur commented on src/bp/lang-server/service/index.ts:
```diff
@@ -41,10 +72,29 @@ export default class LanguageService {
     this._ready = true
+
+    setInterval(() => {
+      const ramInfos = checkEnoughRam(languages.length, this.dim)
```
Extract this in a private method:

```
setInterval(this._runRamCheckJob, ms('1m'))
```

franklevasseur commented on src/bp/lang-server/service/index.ts:
```diff
@@ -3,22 +3,46 @@ import fs from 'fs'
+  return (linear_regression_alpha * dims + linear_regression_beta) / 1000
+}
+
+const checkEnoughRam = (langNb: number, dims: number): RamInfos => {
```
Why is this not a private method?

franklevasseur commented on src/bp/lang-server/service/index.ts:
```diff
@@ -3,22 +3,46 @@ import fs from 'fs'
 // Examples: "scope.en.bpe.model" "bp.en.bpe.model"
 const BPE_MODEL_REGEX = /^(\w+)\.(\w+)\.bpe\.model$/i
 
+const getModelSize = (dims: number) => {
```
This is not a model size getter. It's a model size **estimator**. Something like `estimateModelSize()` is more accurate.

\+ why is this not a private method?

\+ what I meant was to extract alpha and beta in constants with capital case at the top of the file (This is a a widespread convention in cs)

\+ Usually, we don't call a function just alpha or beta... We try to name variables by what they actually represent:

```
alpha => MODEL_MB_PER_DIM ?
beta => MODEL_CONSTANT_MB ?
``` 

How did you estimated this ? Typically, if you've took some data points and estimated a linear regression model, I expect the PR description to include your scientific process (with a graph).

franklevasseur reviewed (changes requested):
The `setInterval` logic is better than checking only when booting, but:

1 - 1 minute might be a long time between checks. From what I observed, during the training of a big bot, memory consumption can increase by few Gbs in what feels like less than 10 seconds 

\+ We don't have to warn the user at every interval. We just have to warn him once to let him know that if the lang server krash, a lack of memory is a highly probable cause.

\+ TBH, a simple `logger.warn` when booting saying something like `"Make sure you have enough memory, if not this program can crash silently."` would have done the job.

2 - The solution I actually expected was something like this:

```ts
process.on('beforeExit', (code) => {
// do your RAM check here
  console.log('Lang server might be crashing because you don't have enough RAM');
});
```

An exit handler that simply explains why the process has exited. I don't know if this will work, but I want you to try this first.

3 - When you open a PR, you're supposed to add a screenshot of the UX result. In your case the screenshot I would expect are the console messages.

I haven't tried your solution yet, but I feel like this is mergeable. 

Try handling a krash in a process exit handler like I said. 

If this doesnt work, check my comments. 

I'll try out your PR. 

Then we'll be good to merge.

Commit:
Merge branch 'master' of github.com:botpress/botpress into ps_warn_ram

Ierezell commented:
Here is the actual log when a too high usage ram is predicted :
```
02/19/2021 12:12:13.943 Launcher ========================================
                                         Botpress Language Server        
                                              Version 1.1.0             
                                              OS linux arch             
                                 ========================================
02/19/2021 12:12:13.946 Launcher authToken: disabled (anyone can query your language server)
02/19/2021 12:12:13.946 Launcher adminToken: disabled (anyone can add, remove or change languages)
02/19/2021 12:12:13.946 Launcher limit: disabled (no protection - anyone can query without limitation)
02/19/2021 12:12:13.946 Launcher Mode: online (languages will be downloaded from https://nyc3.digitaloceanspaces.com/botpress-public/embeddings/index.json)
02/19/2021 12:12:13.946 Launcher Serving 300 language dimensions from /home/pedro/botpress/embeddings
02/19/2021 12:12:13.952 Service Warning, the language server (3.72Gb) may silently crash due to a lack of free memory space. (11.7/15.16)
02/19/2021 12:12:13.952 Service Found Languages: en
02/19/2021 12:12:13.963 Service Loading Embeddings for EN
02/19/2021 12:12:13.964 API Language Server is ready at http://localhost:3100/
02/19/2021 12:12:17.730 Service [EN] Took 3766ms to load 3077mb into RAM (/home/pedro/botpress/embeddings/bp.en.300.bin)
02/19/2021 12:12:17.813 Service [EN] Took 83ms to load 42mb into RAM (/home/pedro/botpress/embeddings/bp.en.bpe.model)
```

Here is the same when the ram is ok : 

```
02/19/2021 12:14:10.959 Launcher ========================================
                                         Botpress Language Server        
                                              Version 1.1.0             
                                              OS linux arch             
                                 ========================================
02/19/2021 12:14:10.961 Launcher authToken: disabled (anyone can query your language server)
02/19/2021 12:14:10.961 Launcher adminToken: disabled (anyone can add, remove or change languages)
02/19/2021 12:14:10.961 Launcher limit: disabled (no protection - anyone can query without limitation)
02/19/2021 12:14:10.962 Launcher Mode: online (languages will be downloaded from https://nyc3.digitaloceanspaces.com/botpress-public/embeddings/index.json)
02/19/2021 12:14:10.962 Launcher Serving 300 language dimensions from /home/pedro/botpress/embeddings
02/19/2021 12:14:10.968 Service Found Languages: en
02/19/2021 12:14:10.979 Service Loading Embeddings for EN
02/19/2021 12:14:10.981 API Language Server is ready at http://localhost:3100/
02/19/2021 12:14:14.000 Service [EN] Took 3020ms to load 3078mb into RAM (/home/pedro/botpress/embeddings/bp.en.300.bin)
02/19/2021 12:14:14.077 Service [EN] Took 76ms to load 42mb into RAM (/home/pedro/botpress/embeddings/bp.en.bpe.model)
```


I removed the `setInterval` and put a `process.on` so user will be informed on boot and on crash only.

Commit:
More concise calls

Commit:
Added ctrl-c  log when  exiting

franklevasseur commented on src/bp/lang-server/service/index.ts:
```diff
@@ -41,6 +82,20 @@ export default class LanguageService {
+      this.checkRam(languages)
+
+      const that = this
+      process.on('SIGINT', function() {
```
leftover? No need for this listener

franklevasseur committed:
minor fixes
